pkg/ingester/incremental_transfer.go

```diff
@@ -0,0 +1,264 @@
+package ingester
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"golang.org/x/net/context"
+)
+
+var (
+	blockedRanges = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_ingester_blocked_ranges",
+		Help: "The current number of ranges that will not accept writes by this ingester.",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(blockedRanges)
+}
+
+// SendChunks accepts chunks from a client and moves them into the local Ingester.
+func (i *Ingester) SendChunks(stream client.Ingester_SendChunksServer) error {
+	i.userStatesMtx.Lock()
+	defer i.userStatesMtx.Unlock()
+
+	fromIngesterID, seriesReceived, err := i.acceptChunks(acceptChunksOptions{
+		UserStates: i.userStates,
+		Stream:     stream,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := stream.SendAndClose(&client.SendChunksResponse{}); err != nil {
+		level.Error(util.Logger).Log("msg", "Error closing SendChunks stream", "from_ingester", fromIngesterID, "err", err)
+		return err
+	}
+
+	// Target ingester may not have had any streams to send.
+	if seriesReceived == 0 {
+		return nil
+	} else if fromIngesterID == "" {
+		level.Error(util.Logger).Log("msg", "received TransferChunks request with no ID from ingester")
+		return fmt.Errorf("no ingester id")
+	}
+
+	return nil
+}
+
+// GetChunks accepts a get request from a client and sends all chunks from the serving ingester
+// that fall within the given range to the client.
+func (i *Ingester) GetChunks(req *client.GetChunksRequest, stream client.Ingester_GetChunksServer) error {
+	userStatesCopy := i.userStates.cp()
+	if len(userStatesCopy) == 0 {
+		level.Info(util.Logger).Log("msg", "nothing to transfer")
+		return nil
+	}
+
+	// Block the ranges locally so we don't receive traffic for it anymore.
+	// Will be later unblocked by CleanupTransfer.
+	if err := i.BlockRanges(stream.Context(), req.Ranges); err != nil {
```

Comment:
We don't call `UnblockRanges` at the end of `GetChunks` because we want writes to continue to be rejected until the joining ingester has successfully added its token into the ring. `CleanupTransfer` is used to protect against spillover. The automatic unblock is used as a failsafe in case the joining ingester fails or crashes: writes will continue to go to the blocked ranges and be rejected. We want them to succeed eventually, so we let them unblock themselves after a timeout.

Reply:
I don't like spawning of new goroutine for each BlockRanges, would prefer single goroutine that expires blocked ranges, if any. To make what you write work properly, should joining ingester keep extending those blockages until it is ready to unblock it all?

Reply:
I'm not sure how I feel about a permanent goroutine that waits for blocked ranges and sees if they need to be removed after an expiration period. It would also mean we'd have to introduce a new `TimedRange` (or something). I'll put some thought into it. 

> To make what you write work properly, should joining ingester keep extending those blockages until it is ready to unblock it all?

Implying that this doesn't work properly? 🙂 I feel that it's up to operators to configure the timeout to be longer than a join takes. Adding in extensions to the blocked range timeout seems like it would be a lot of unnecessary logic.

Reply:
But how can operator guess how long it will take? What if he guesses wrongly? On the other hand, ingester knows when it's ready, and if it's not but still running, it can extend the timeout.

Reply:
> I'm not sure how I feel about a permanent goroutine that waits for blocked ranges and sees if they need to be removed after an expiration period.

It can be spawned only when needed and not running yet. If goroutine unblocks everything, it can die again. (ok, that's somewhat more complex than what I had in mind. I think single goroutine running all the time and checking timeouts would be just fine).

Reply:
> But how can operator guess how long it will take? What if he guesses wrongly?

It should just be set to a large value, overshooting shouldn't be a practical concern. The default timeout should be covering it for most clusters, and if you're running a giant cluster, shutdown duration is a known metric.

Reply:
Or it can be one less thing to worry about with little extra code? ;-) But I see your point. Basically you would just use large enough timeout to be sure it's ok.